File input requirement aka AvatarUploader

isacoder opened the pull request:
### Requirements

[Bugherd21](https://www.bugherd.com/projects/216308/tasks/21)
[New] File type input

Design reference
https://app.zeplin.io/project/5fb77e662d5c19b7d12dca42/screen/5fbd0fa72672e75d9f5a6de1
https://zpl.io/2p5DJjj

### Implementation
4 new components expected in this implementation

- **FileInputButton** :  Normal button with sizes and design capabilities that has a callback to retrieve the selected files. (Input type file props can be set).

- **DropArea**: Area with 100% width and Height as a prop that has a dropCallback to get the selected files, the background is transparent so you can add it to positioned absolute to any container and style any background needed.

- **CropArea** : Specified the canvas size and crop size  and imgUrl returns a new imgUrl with the selected area.

- **AvatarUploader** : Component that uses the previous 3 components to get a file, edit it and display it.

~~Made a useCrop hook for the Crop Area
This is my first attempt with useReducer, feedback will be appreciated.
I reviewed the two implementations that we have in the code, but wasn't sure how to use mo~~

Took some liberties with the design of the CropTool:
- Since canvas can have a different size than the image, when the sizes are different any space left will have a background.

~~Crop tool has some flickering similar to the one I saw in the slider :/ not sure if this is only happening in dev and will not look in production. I'll be try to look more into it in a next version.~~

I moved all this 4 components to the Folder Story of File management

### Screenshots

**FileInputButton**
<img width="483" alt="Screen Shot 2021-05-11 at 10 26 39" src="https://user-images.githubusercontent.com/10409078/117744296-614e3380-b243-11eb-8af3-98eed98842ed.png">

**DropArea**
<img width="676" alt="Screen Shot 2021-05-11 at 10 28 30" src="https://user-images.githubusercontent.com/10409078/117744434-a3777500-b243-11eb-9cfe-4b8f649b5566.png">


**CropArea**
<img width="989" alt="Screen Shot 2021-05-11 at 10 29 26" src="https://user-images.githubusercontent.com/10409078/117744495-c43fca80-b243-11eb-822c-ae97f4de8fb9.png">

**AvatarUploader**
<img width="519" alt="Screen Shot 2021-05-11 at 10 29 57" src="https://user-images.githubusercontent.com/10409078/117744529-d6216d80-b243-11eb-9bd8-c3116065d69a.png">

atomworks commented on src/Form/organisms/AvatarUploader.tsx:
```diff
@@ -0,0 +1,156 @@
+import Label from '../atoms/Label';
+import { isValidImage } from '../../helpers';
+
+const PHOTO_HEGHT = `150px`;
```
Typo on height

isacoder replied:
updated

atomworks commented on src/Form/organisms/AvatarUploader.tsx:
```diff
@@ -0,0 +1,156 @@
+            imgUrl={avatarImg}
+            onCrop={handleCrop}
+            onClose={handleCropClose}
+            canvasHeight={490}
+            canvasWidth={460}
+            cropHeight={360}
+            cropWidth={360}
```
Are these just defaults? Wasn't expecting to see hard coded values here for `CropTool` so it seems like these could be a little problematic when you have unexpected aspect ratios. Could this cause any issues?

isacoder replied:
The reason behind the Croptool getting this values look "hard coded" is that the parent of the croptool is the one that controls the size of the canvas and the size of the selection. (although if the size of the selection is is bigger it will be ajusted to the canvas).  

From the examples I reviewed the tools allow the user of the UI library to provide this values, although sometimes allow to receive an aspect radio instead of those and the UI plugin would calculate the canvas size for them. Maybe we can have a call to better understand the expectations.

atomworks commented on src/Form/molecules/CropTool.tsx:
```diff
@@ -0,0 +1,399 @@
+  left: -5px;
+`;
+
+function drawImgValues(img: HTMLImageElement, canvasHeight: number, canvasWidth: number) : IDrawArea {
```
Better to use arrow format functions

isacoder replied:
updated

atomworks reviewed (changes requested):
Looks good on the whole. About to do a check in on the usability side and will report on that separately.

The feedback points are mostly related to typos, function declarations and simpler things. There are a couple of more important notes or questions but nothing that demands any major rework.

It seems like things could be broken down a little more in to more manageable / easy to read files / components but I don't think that refactor is critical at all at this point.

atomworks commented:
Just to add as I've review the usability too. As you were concerned about, the redraw issue causing re-rendering of other elements is problematic. Perhaps for a beta release we can provide it for people to use but afterwards we should follow up as quickly as possible to resolve that.

One small bug I noticed too is that the cropped image in the Avatar Uploader doesn't have the correct aspect ration, it gets a bit squashed. I think for this it's fine to have part of the image hidden if it goes beyond the bounds of the container. If it's centered than for a profile photo use case it should still perform the task as needed.

Other than that though it's all looking great. We might want to expand out some example usages of the callbacks on the Drop Area and Input File Button components but I think we we can do that later and perhaps as part of an expansion of that the designed set of components.

isacoder commented on src/hooks/useCrop.ts:
```diff
@@ -0,0 +1,377 @@
+import { clamp } from '../helpers/index';
+
+// https://developer.mozilla.org/en-US/docs/Web/CSS/cursor //
+export type ICursorStyles = (
```
Well in the case of type vs interfaces, both look pretty similar but they have different capabilities.
It's not that one is new and the other is old rather than depends in the intention.

Typescript will throw an error if a type is decelerated and the same name type already exist but allows to have multiple declarations of interfaces with the same name and merges them.

https://thisthat.dev/interface-vs-type/

Although even the page recommends the usage of interface over types, there are scenarios where type is necessary.  In this case specific case, since I'm declaring specific "string" values I need it to use a type to declare them. This case is similar to design and sizes in buttons.

I haven't seen it used much in the base code, but I guess I could try enums for this case.
https://blog.logrocket.com/writing-readable-code-with-typescript-enums-a84864f340e9/

Just as a comment aside,  another occasion to use types is when you need to join more than one interface, because is not posible to extend more than one, but you can join more than one.
https://blog.logrocket.com/types-vs-interfaces-in-typescript/

By no means I'm an expert on the subject and  I've seen on internet that there's not a great agreement about it's usage, just hundreds of examples of how people use them.

isacoder commented on src/hooks/useCrop.ts:
```diff
@@ -0,0 +1,377 @@
+      return {...state, cursorStyle: action.data.cursorStyle};
+
+    case START_RESIZE:
+      {
```
This is in other to be able to declare a variable inside the case, if not it will complain.

https://medium.com/@e_himmelfarb/use-curly-braces-with-es6-let-and-const-in-switch-blocks-react-redux-reducers-c0b01b37d748

isacoder commented on src/hooks/useCrop.ts:
```diff
@@ -0,0 +1,377 @@
+  }
+}
+
+export function useCrop(initialValues = defaultCropValues) {
```
updated!

isacoder commented on src/Form/atoms/DropArea.tsx:
```diff
@@ -0,0 +1,77 @@
+  position: relative;
+`;
+
+const DradAndDrop = styled.div<{inDropZone: boolean}>`
```
Updated

isacoder commented on src/Form/atoms/DropArea.tsx:
```diff
@@ -0,0 +1,77 @@
+},[dropCallback]);
+
+  return(
+    <Container height={height}>
```
updated!

isacoder commented on src/Form/atoms/InputFileButton.tsx:
```diff
@@ -0,0 +1,59 @@
+      }
+  },[callback]);
+
+  const fowardClick = useCallback(() => {
```
updated

isacoder commented on src/helpers/index.tsx:
```diff
@@ -21,4 +21,23 @@ export const getTextTimeUnit = (value: number, unit : ITimeUnit) => {
+    return dataType.split('/')[1].split(';')[0];
+  }
+  return 'image/jpeg';
+};
```
Well this is based on the Mime Types

https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types

I'm using this helper to determinate the type of the image base and put the same in the return new image created.

Since I'm drawing the image in the canvas. If I draw an image that was a png  I wan to crop and get a png. If I receive a jpeg, I want to crop and obtain a jpeg.. if I receive any other kind of image (but it has to be an image) I will generate a jpeg.

https://developer.mozilla.org/en-US/docs/Web/API/HTMLCanvasElement/toDataURL

I kind of remembered that you mentioned that we were limiting to jpeg and png for the AvatarUploader, but wasn't sure it has to be the same for the Croptool or if we can receive other types
https://developer.mozilla.org/en-US/docs/Web/Media/Formats/Image_types

In any case I'm validating that we receive from the drop file and the input file a valid image jpeg, png.. I could add more if is expected.

```js
const isValidImage = (file: File) => {
  const acceptedImageTypes = ['image/jpeg', 'image/png'];

  return file && acceptedImageTypes.includes(file['type']);
};
```

isacoder commented:
Hello!
I have updated the changes and this PR is ready for review again.

- I corrected all the typos and added the spell checker plugging to my VS code to prevent them 🙈 
- Updated all functions to arrow functions as requested.
- Fixed the re-rendering  by updating the Croptool selection functionality, it will no longer rely on state to repaint the selection. I went full vanilla for this implementation and used the cropRef to update the dimensions directly in the Dom.
Because of that removed de useReduce.
- Abstracted the UI part of the selection to CropArea and the functionality to a cropHelper.

atomworks reviewed (changes requested):
There's a few small changes I'd like to make before we merge this. 

- Bug: If you select a file and in Avatar Uploader and then replace the file with the same one it doesn’t bring up the crop.
- In the Crop Tool story, can we set resizable and leave aspect blank by default in the Knobs. I think that will better demonstrate the flexibility of the tool, especially as we have these switched on in the avatar uploader.
- Also, if we’re able to set the default crop on the image to be the largest size possible for the aspect ratio with the crop area centered in the image, that would be great too

Those small things aside, really happy with this. Looks great and is running really smoothly now!

isacoder commented:
Hello finished the required updates.

atomworks reviewed (approved):
Thanks for making those changes. All looks great!

@JoshLipps I think you're on reviewers for this. If you're happy, feel free to get it approved and merged.
